Fehler beim speichern von Fehlern in xtcasError.

Wenn ich das log lese, sieht es wie ein Fehler aus:
* [ ] Falls tatsächlich der Stacktrace in der Datenbank gespeichert wird,
wäre es gut, wenn man statt `CAS : Execute` sowas wie `CAS : Saving stacktrace` loggt.
Sonst wirkt es für mich persönlich so,
als würde beim SQL-Ausführen ein Fehler auftreten.
* [ ] Falls der Stacktrace nicht abgespeichert wird,
sollte dieser auch nicht bei diesem SQL-Statement geloggt werden.
Man muss dabei aber sicherstellen,
dass die Stacktraces von Exceptions dabei immernoch geloggt werden.
Diese Stacktraces sollten aber in einer getrennten Log-Meldung erscheinen
und nicht in der selben Meldung, wo die Message des Fehlers in der Datenbank gespeichert wird.

```
2021-12-21T12:00:58.140 CAS : Execute : INSERT INTO xtcasError (Username, ErrorMessage, Date) VALUES (?,?,?) with values: Minova.Support-o@mabanaft.intra, java.lang.RuntimeException: org.springframework.web.client.HttpServerErrorException$InternalServerError: 500 : [{"timestamp":"2021-12-21T12:00:58.133+0000","status":500,"error":"Internal Server Error","message":"java.lang.RuntimeException: Report found! Please delete the exisiting report first!","path":"/createReport"}], 2021-12-21 12:00:58.140367/naero.minova.core.application.system.domain.ProcedureException: java.lang.RuntimeException: org.springframework.web.client.HttpServerErrorException$InternalServerError: 500 : [{"timestamp":"2021-12-21T12:00:58.133+0000","status":500,"error":"Internal Server Error","message":"java.lang.RuntimeException: Report found! Please delete the exisiting report first!","path":"/createReport"}]
	at aero.minova.core.application.system.controller.SqlProcedureController.executeProcedure(SqlProcedureController.java:141)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:879)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:126)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:90)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:118)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:158)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilterInternal(BasicAuthenticationFilter.java:204)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:92)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:367)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1639)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.RuntimeException: org.springframework.web.client.HttpServerErrorException$InternalServerError: 500 : [{"timestamp":"2021-12-21T12:00:58.133+0000","status":500,"error":"Internal Server Error","message":"java.lang.RuntimeException: Report found! Please delete the exisiting report first!","path":"/createReport"}]
	at aero.minova.birt.report.BirtReport.createBirtReport(BirtReport.java:249)
	at aero.minova.birt.report.BirtReport.lambda$setup$3(BirtReport.java:111)
	at aero.minova.core.application.system.controller.SqlProcedureController.executeProcedure(SqlProcedureController.java:130)
	... 86 more
Caused by: org.springframework.web.client.HttpServerErrorException$InternalServerError: 500 : [{"timestamp":"2021-12-21T12:00:58.133+0000","status":500,"error":"Internal Server Error","message":"java.lang.RuntimeException: Report found! Please delete the exisiting report first!","path":"/createReport"}]
	at org.springframework.web.client.HttpServerErrorException.create(HttpServerErrorException.java:100)
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:172)
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:112)
	at org.springframework.web.client.ResponseErrorHandler.handleError(ResponseErrorHandler.java:63)
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:782)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:740)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:674)
	at org.springframework.web.client.RestTemplate.postForEntity(RestTemplate.java:449)
	at aero.minova.birt.report.BirtReport.createBirtReport(BirtReport.java:244)
	... 88 more
```

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -5,6 +5,8 @@ Das Format basiert auf link:https://keepachangelog.com/en/1.0.0[Keep a Changelog
 
 ## [Unreleased]
 * API für Extensions, welche Nachrichten an einen Dienst schicken möchten.
+* #274 Fehlermeldungen, welche nicht mit 'msg.' beginnen, verarbeitbar machen.
+* #291 Doku für Logs hinzugefügt
 
 
 ## [12.46.0] - 2022-06-09

--- a/service/doc/adoc/index.adoc
+++ b/service/doc/adoc/index.adoc
@@ -6,6 +6,7 @@
 
 * xref:extensions.adoc#[Erweiterungen]
 * xref:properties.adoc#[Properties]
+* xref:logs.adoc#[Logging und Logs]
 * xref:./../../../setup/doc/adoc/setup.adoc#[Setup Service]
 * xref:./servicenotifier.adoc#[ServiceNotifier Service]
 * xref:queueservice.adoc#[Queue Service]

--- a/service/doc/adoc/logs.adoc
+++ b/service/doc/adoc/logs.adoc
@@ -4,7 +4,7 @@ Die Logs des CAS werden im logs-Ordner im root-Verzeichnis abgelegt.
 
 Es gibt sieben verschiedene Logs:
 
-* *errorLog*: Loggt alle auftretenden Fehlermeldungen.
+* *errorLog*: Loggt alle auftretenden Fehlermeldungen und den dazugehörigen Stacktrace.
 * *privilegeLog*: Loggt alle Privilegienanfragen, welche das CAS durchführt.
 * *springLog*: Loggt alle SQL-Anfragen vom CAS an die Datenbank.
 * *userLog*: Loggt alle Anfragen der User an das CAS.

--- a/service/doc/adoc/logs.adoc
+++ b/service/doc/adoc/logs.adoc
@@ -1,0 +1,21 @@
+== Logging und Logs
+
+Die Logs des CAS werden im logs-Ordner im root-Verzeichnis abgelegt.
+
+Es gibt sieben verschiedene Logs:
+
+* *errorLog*: Loggt alle auftretenden Fehlermeldungen.
+* *privilegeLog*: Loggt alle Privilegienanfragen, welche das CAS durchführt.
+* *springLog*: Loggt alle SQL-Anfragen vom CAS an die Datenbank.
+* *userLog*: Loggt alle Anfragen der User an das CAS.
+* *filesLog*: Loggt alle File-Anfragen von Usern und alle Hash- und Zip-Operationen des CAS.
+* *setupLog*: Loggt alle Operationen, welche das CAS beim Setup durchführt.
+* *queueServiceLog*: Loggt jede Aktivität des QueueService.
+
+Nachdem ein neuer Tag anbricht, verschiebt das CAS die 'alten' Logs automatisch in einen 'archived'-Ordner innerhalb des logs-Ordners.
+Somit gibt es für immer genau *ein Log jeder Art pro Tag*.
+
+Das Log-Pattern und das Namens-Pattern der Logs wird durch die logback-spring.xml im src/main/resources bestimmt.
+Dort werden die Logfiles auch mit der CustomLogger-Klasse verbunden.
+
+Das Pattern der Logs im Allgemeinen ist immer: *yyyy-MM-dd'T'HH:mm:ss.SSS Benutzer:*

--- a/service/doc/adoc/logs.adoc
+++ b/service/doc/adoc/logs.adoc
@@ -1,6 +1,6 @@
 == Logging und Logs
 
-Die Logs des CAS werden im logs-Ordner im root-Verzeichnis abgelegt.
+Die Logs des CAS werden im APP_LOG_ROOT-Verzeichnis abgelegt, welches als Default einen logs-Ordner im Root-Verzeichnis des CAS Dienstes anlegt.
 
 Es gibt sieben verschiedene Logs:
 

--- a/service/src/main/java/aero/minova/cas/ControllerExceptionHandler.java
+++ b/service/src/main/java/aero/minova/cas/ControllerExceptionHandler.java
@@ -123,7 +123,9 @@ public class ControllerExceptionHandler extends ResponseEntityExceptionHandler {
 		}
 
 		// Alles vor 'msg.' wegschmeißen.
-		errorMessage = errorMessage.substring(errorMessage.indexOf("msg."));
+		if (errorMessage.contains("msg.")) {
+			errorMessage = errorMessage.substring(errorMessage.indexOf("msg."));
+		}
 
 		/*
 		 * Es gibt zwei Fehlermeldungsformate: 1. 'ADO | 25 | msg.sql.51103 @p tUnit.Description.16 @s kg | Beipieltext' 2. 'msg.PrivilegeError %tBeispiel'

--- a/service/src/main/java/aero/minova/cas/ControllerExceptionHandler.java
+++ b/service/src/main/java/aero/minova/cas/ControllerExceptionHandler.java
@@ -317,8 +317,8 @@ public class ControllerExceptionHandler extends ResponseEntityExceptionHandler {
 				StringWriter sw = new StringWriter();
 				PrintWriter pw = new PrintWriter(sw);
 				e.printStackTrace(pw);
-				logger.info(
-						"CAS : Execute : " + errorStatement + " with values: " + username + ", " + e.getMessage() + ", " + timeOfError + "/n" + sw.toString());
+				logger.info("CAS : Saving stacktrace : " + errorStatement + " with values: " + username + ", " + e.getMessage() + ", " + timeOfError + "/n"
+						+ sw.toString());
 			}
 			callableErrorStatement.executeUpdate();
 			connection.commit();

--- a/service/src/main/java/aero/minova/cas/ControllerExceptionHandler.java
+++ b/service/src/main/java/aero/minova/cas/ControllerExceptionHandler.java
@@ -317,8 +317,9 @@ public class ControllerExceptionHandler extends ResponseEntityExceptionHandler {
 				StringWriter sw = new StringWriter();
 				PrintWriter pw = new PrintWriter(sw);
 				e.printStackTrace(pw);
-				logger.info("CAS : Saving stacktrace : " + errorStatement + " with values: " + username + ", " + e.getMessage() + ", " + timeOfError + "/n"
-						+ sw.toString());
+				customLogger.logSql("CAS : Execute : " + errorStatement + " with values: " + username + ", " + e.getMessage() + ", " + timeOfError);
+				// Der Stacktrace wird nicht in der Datenbank gespeichert, da das Feld einfach viel zu lang ist. Deswegen geben wir ihn im ErrorLog aus.
+				customLogger.errorLogger.info("CAS: Showing Stacktrace : " + sw.toString());
 			}
 			callableErrorStatement.executeUpdate();
 			connection.commit();
@@ -326,7 +327,7 @@ public class ControllerExceptionHandler extends ResponseEntityExceptionHandler {
 			StringWriter sw = new StringWriter();
 			PrintWriter pw = new PrintWriter(sw);
 			e1.printStackTrace(pw);
-			customLogger.errorLogger.error("CAS : Error could not be saved in database." + "/n" + sw.toString());
+			customLogger.logError("CAS : Error could not be saved in database." + "/n" + sw.toString(), e1);
 		} finally {
 			systemDatabase.freeUpConnection(connection);
 		}

